File: sdk/js/src/schemas/null.ts
import type { ParseContext, SchemaNode } from "../types.js";
import { BaseSchema } from "./base.js";
import { ISSUE_CODES } from "../issue-codes.js";
import { describeType } from "../util.js";

export class NullSchema extends BaseSchema<null, null> {
  _validate(input: unknown, ctx: ParseContext): unknown {
    if (input !== null) {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_TYPE,
        message: `Expected null, received ${describeType(input)}`,
        path: [...ctx.path],
        expected: "null",
        received: describeType(input),
      });
      return undefined;
    }
    return null;
  }

  _toNode(): SchemaNode {
    const node: SchemaNode = { kind: "null" } as SchemaNode;
    this._addDefault(node);
    return node;
  }
}

File: sdk/python/src/anyvali/schemas/null.py
"""Null schema."""

from __future__ import annotations

from typing import Any

from ..issue_codes import INVALID_TYPE
from .base import BaseSchema, ValidationContext, _anyvali_type_name


class NullSchema(BaseSchema[None]):
    """Schema that only accepts None/null."""

    def __init__(self) -> None:
        super().__init__()

    def _accepts_none(self) -> bool:
        return True

    def _validate(self, input: Any, ctx: ValidationContext) -> Any:
        if input is not None:
            received = _anyvali_type_name(input)
            ctx.add_issue(INVALID_TYPE, f"Expected null, received {received}", expected="null", received=received)
            return None
        return None

    def _to_node(self) -> dict[str, Any]:
        return self._add_common_node_fields({"kind": "null"})
